compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -353,49 +353,76 @@ fn negative_impl<'cx, 'tcx>(
     })
 }
 
+/// Try to prove that a negative impl exist for the given obligation and their super predicates.
+#[instrument(level = "debug", skip(selcx))]
 fn negative_impl_exists<'cx, 'tcx>(
     selcx: &SelectionContext<'cx, 'tcx>,
     param_env: ty::ParamEnv<'tcx>,
     region_context: DefId,
     o: &PredicateObligation<'tcx>,
 ) -> bool {
     let infcx = &selcx.infcx().fork();
+
+    if resolve_negative_obligation(infcx, param_env, region_context, o) {
+        return true;
+    }
+
+    // Try to prove a negative obligation exist for super predicates
+    // FIXME we should implement an `assemble_neg_candidates` fn
```

Comment:
Why?

Reply:
So, my take is this:

* We defined (in this PR) that `T: !Foo` means "there is a negative impl for `Foo`". But I think the *right* meaning is more "`T` can never implement `Foo`" -- that is the question coherence is interested in.
* Right now, coherence is using the more narrow question to answer the bigger one. Since it's the only person asking the question, that's fine, but if we had other bits of code also trying to prove negative predicates, I would prefer if we redefined what negative predicates mean.

Reply:
That said, I'm tempted to say we should remove this FIXME, I don't think it will do anyone any good and it's kind of just going to confuse people.
```suggestion
```